### DISCUSSION TITLE
Is there a way to run tgcf on a raspberry pi?

### DISCUSSION THREAD

**Discussion by ghost** [aahnik/tgcf#150] [Q&A] [answered]
I bought a raspberry pi to run tgcf, but it doesn't seem to work. 

```shell
Docker : standard_init_linux.go:207: exec user process caused "exec format error"
pip install tgcf : 'could not find a version that satisfies the requirement tgcf'
```

Is there any way to run tgcf on the Raspberry Pi?

**Comment by aahnik** [comment]
I think `tgcf` can run on any Linux environment.

Please tell me what command did you run to install `tgcf`? Have you used the docker method or the `pip` method? Your question is not clear.

For using tgcf on docker, `docker` must be installed on the system. Run `docker --version ` to check. Read how to [install via docker](https://github.com/aahnik/tgcf/wiki/Install-and-run-using-docker) here.

For using the `pip` installation, python must be installed. Is python installed in your system?
You may run these to check:

```shell
python --version
pip --version
```

**Reply by ghost** [reply]
The device I used is 'Raspberry Pi 4b', and I tried using Docker.

The command I entered is as follows.


`sudo apt update`
`sudo apt full-upgrade`
`sudo reboot`

`sudo apt install docker.io`

`docker --version`  
 `docker version 18.0.9.1 build 4c52b90`

`sudo usermod -aG docker $USER`
`sudo reboot`

`docker pull aahnik/tgcf`
`status: Image is up to date for aahnik/tgcf:latest`

`docker run -v /home/pi/tgcf/tgcf.config.yml:/app/tgcf.config.yml --env-file /home/pi/tgcf/.env -it aahnik/tgcf`
(I have created the '.env' file and the 'tgcf.config.yml' file in the /home/pi/tgcf path.)

And, the following error is printed out.

`standard_init_linux.go:207: exec user process caused "exec format error"`

With the same command, I checked that it worked well with docker for Windows. I don't know what the problem is.
Please help me!



--------------------------------------

Also, I applied the solution of the link below. 
https://www.dahlen.org/2019/06/docker-und-raspbian-auf-einem-raspberry-pi/

As a result of applying the above solution, the docker version became 20.10.6, build 370c289.
And there was a slight difference! The error message below is printed out.

`WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm/v7) and no specific standard_init_linux.go:219: exec user process caused: exec format error`

**Comment by aahnik** [comment]
thank you for a detailed response. The last line is helpful.

> The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm/v7) and no specific standard_init_linux.go:219: exec user process caused: exec format error

The docker image published at `aahnik/tgcf` is not supported for raspberry pi.

If you have knowledge of docker, you can build a docker image for raspberry pi, and publish it to docker hub, or upload the docker image from your computer to raspberry pi.

----------------

Simpler alternative for you would be to :
- install python and pip in your raspberry pi
- `pip install tgcf`

**Comment by aahnik** [comment]
# A dockerfile for rasbberry pi

You can build a docker image for raspberry pi.

- You need docker installed in your main computer.

Step 1. Create a folder and a file `Dockerfile` inside it.

Step 2. Put this in the Dockerfile

```Dockerfile
FROM arm64v8/python:3.9 
# search docker hub for the base image for the architecture of your hardware

WORKDIR /app

RUN pip install --upgrade tgcf

CMD tgcf
```

Step 3. Open your terminal inside the folder that contains the Dockerfile, and run this

```shell
docker build -t tgcf-rspi .
```

The name `tgcf-rspi` can be anything else you want.

Once the image has been build, you can upload it to DockerHub, in your account. Then you may pull and run that in your raspberry pi.

**Reply by ghost** [reply]
Wow, thank you. I'm going to try this right now!

**Reply by ghost** [reply]
I had a little problem trying this. I will try to solve it.

**Comment by ghost** [comment]
I've been trying to run tgcf on Raspberry Pi for a few days, but unfortunately I've given it up.
However, I will write down the process I have tried here because I think it might be a little helpful for someone who will try the same as me.

First, how to create a Dockerfile.

```
FROM python:3.7
WORKDIR /app
RUN pip install --upgrade tgcf
CMD tgcf
```


Then run the `docker build --platform linux/arm/v7 -t <tag> <location_dockerfile>` command. Then the compilation will be made for Raspberry Pi. Then upload the created image to Docker and pull it from raspberry pi.

The problem is, Tgcf needs **Python 3.9**. If you use a file compiled with **Python3.7**, when you type `docker run --upgrade tgcf`, **version 0.0.1 of tgcf** is installed, which does not work properly.



```
FROM python:3.9
WORKDIR /app
RUN pip install --upgrade tgcf
CMD tgcf
```

To solve this problem, I changed the contents of Dockerfile to **Python:3.9**. However, the compile is not working properly and shows an error. This seems to be a problem related to 'Pillow', but I have not found a solution.
`ERROR: Failed building wheel for Pillow`

Also, Raspberry Pie OS does not officially support **Python3.9** yet. Of course, I can install version 3.9 in Raspberry Pi, but I didn't go that far. Because, as I said before, I failed to cross compile.

I have little knowledge of programming. (I'm just starting to learn) So I tried to figure it out by rule of thumb, but unfortunately it didn't work out. For the time being, I'll just use andriod. But I hope someone will solve this problem well. Go for it! 
And Thank you @aahnik for giving me directions on the solution.

**Reply by aahnik** [reply]
hi @comienzo2093, thanks for documenting.

**Reply by aahnik** [reply]
hi @comienzo2093 try this Dockerfile

```Dockerfile
FROM python:3.9

WORKDIR /app

RUN apt-get update && apt-get upgrade -y

RUN pip install --upgrade pip tgcf

CMD tgcf
```

From the folder this Dockerfile is stored, 

```shell
docker build --platform linux/arm/v7 -t tgcf:raspberry .
```

**Reply by ghost** [reply]
> hi @comienzo2093 try this Dockerfile
> 
> ```dockerfile
> FROM python:3.9
> 
> WORKDIR /app
> 
> RUN apt-get update && apt-get upgrade -y
> 
> RUN pip install --upgrade pip tgcf
> 
> CMD tgcf
> ```
> 
> From the folder this Dockerfile is stored,
> 
> ```shell
> docker build --platform linux/arm/v7 -t tgcf:raspberry .
> ```

Thank you for the information. However, even if I try like this, It shows the same error.
`ERROR: Failed building wheel for Pillow`
I just succeeded in compiling with the `docker build --platform linux/arm64/v8 -tgcf:raspberry .` command, And it runs normally on **2021-03-04-raspios-buster-arm64(beta)**. I will test a little bit and update the document. Thank you!

**Comment by ghost** [comment]
Thanks to @aahnik help, I succeeded in running tgcf on Raspberry Pi 4b. It's been working properly for two days now. Here's how:



1. First, First, you need to make an image for raspberry pie.


Dockerfile

```
FROM python:3.9

WORKDIR /app

RUN apt-get update && apt-get upgrade -y

RUN python3 -m pip install --upgrade pip tgcf

CMD tgcf
```

From the folder this Dockerfile is stored,

`docker build --platform linux/arm64/v8 -t tgcf:raspberry .`

If the image is made, `push` it.




2. You must install a 64-bit version of Raspberry Pi OS. This is because an error occurs during arm32v7 compilation due to a problem with 'Pliow'.

https://downloads.raspberrypi.org/raspios_arm64/images/


3. Install the latest version of python on Raspberry Pi. Raspberry Pi has 3.7 versions, but tgcf supports 3.9. You can install it by following the guide in the link below. Currently, 3.9.5 is the latest version, and I installed it.

https://itheo.nl/install-python-3-9-on-raspberry-pi


4. `Install` the docker on Raspberry Pi, `pull` the compiled image, and `run` it using command.

If you need more help, leave a reply. I'll help you with what I can.

**Reply by ghost** [reply]
I solved the problem of 'Pillow' installation failure.

`pip install --upgrade pip`
`pip install --upgrade pillow`
`sudo apt-get install libtiff4-dev libjpeg8-dev zlib1g-dev libfreetype6-dev liblcms2-dev libwebp-dev tcl8.5-dev tk8.5-dev python-tk`

https://stackoverflow.com/questions/34631806/fail-during-installation-of-pillow-python-module-in-linux

If you want, you can compile tgcf(arm32v7) on docker bullidx or raspberry pi and then try running it.

I already use the 64-bit version of raspberry pi os, and tgcf(arm64v8) works fine. However, 64-bit raspi os is still beta, and there may be some problems with its use. Therefore, I add this comment for those who want to use 32-bit raspbian.

**Comment by aahnik** [comment, marked as answer]
Make sure you have git and docker installed on your Raspberry pi.

Run these to check.

```shell
git --version
docker --version
```

You need a host of arm architecture to build a docker image for `arm`.

To see the architecture of the host machine, 

```shell
uname -m
````


Follow the steps on your raspberry pi.

- Clone the tgcf repo and move into it. 
    ```shell
      git clone https://github.com/aahnik/tgcf.git && cd tgcf
    ````

- Build the docker image
    ```shell
    docker build -t tgcf:rasp .
    ```

- Come out of the `tgcf` folder, and create a new folder to store your configuration.

- Run `tgcf`
    ```shell
   # make sure to be in the directory which has .env and tgcf.config.yml
    docker run -v /path/to/tgcf.config.yml:/app/tgcf.config.yml -d --env-file .env tgcf:rasp
    ```

**Reply by ghost** [reply]
This method works great and does not require a main computer. Simple! 
I've had a few days of raspberry pie trouble, but now it's all worked out. Thank you.
